internal/logsink/logsink.go

```diff
@@ -0,0 +1,263 @@
+
+	in  chan loggo.Entry
+	out chan []logRecord
+
+	pool sync.Pool
+}
+
+// NewLogSink creates a new log sink that writes log messages to a file. There
+// can only be one writer writing to the same file at a time, otherwise bytes
+// will be written to the file in an interleaved manner (junk data).
+// LogSink writer will write log messages as JSON objects, one per line, even
+// if the log message is multiline.
+func NewLogSink(writer io.Writer, batchSize int, flushInterval time.Duration) *LogSink {
+	return newLogSink(writer, batchSize, flushInterval, nil)
+}
+
+// newLogSink creates a new log sink that writes log messages to a file.
+func newLogSink(writer io.Writer, batchSize int, flushInterval time.Duration, internalStates chan string) *LogSink {
+	w := &LogSink{
+		internalStates: internalStates,
+
+		writer: writer,
+
+		batchSize:     batchSize,
+		flushInterval: flushInterval,
+
+		in:  make(chan loggo.Entry, batchSize),
+		out: make(chan []logRecord, batchSize),
+
+		pool: sync.Pool{
+			New: func() any {
+				return logRecord{}
+			},
+		},
+	}
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Write sends a new log message to the writer.
+// This implements the loggo.Writer interface.
+func (w *LogSink) Write(entry loggo.Entry) {
+	select {
+	case <-w.tomb.Dying():
+		return
+	case w.in <- entry:
+	}
+}
+
+// Kill stops the writer.
+func (w *LogSink) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait blocks until the writer has stopped.
+func (w *LogSink) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *LogSink) loop() error {
+	w.tomb.Go(func() error {
+		buffer := new(bytes.Buffer)
+		encoder := json.NewEncoder(buffer)
+
+		for {
+			select {
+			case <-w.tomb.Dying():
+				return tomb.ErrDying
+			case records := <-w.out:
+				if err := w.write(buffer, encoder, records); err != nil {
+					return err
+				}
+			}
+		}
+	})
+
+	timer := time.NewTimer(w.flushInterval)
+	entries := make([]loggo.Entry, 0, w.batchSize)
+
+	// Tick-toc the in and out channels, to ensure that we can send the batch
+	// of log messages to the underlying writer.
+	in := w.in
+	var out chan []logRecord
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			if len(entries) > 0 {
+				// To prevent a race, we need our own buffer and encoder to
+				// write the remaining log messages that weren't written yet.
+				buffer := new(bytes.Buffer)
+				encoder := json.NewEncoder(buffer)
+				if err := w.write(buffer, encoder, w.records(entries)); err != nil {
+					return err
+				}
+			}
+			return tomb.ErrDying
+
+		case entry := <-in:
+			// Consume log entries until we have a full batch.
+			entries = append(entries, entry)
+			if len(entries) < w.batchSize {
+				continue
+			}
+
+			in = nil
+			out = w.out
+
+			timer.Reset(w.flushInterval)
+
+		case <-timer.C:
+			if len(entries) == 0 {
+				continue
+			}
+
+			in = nil
+			out = w.out
+
+			timer.Reset(w.flushInterval)
+
+			w.reportInternalState(stateTicked)
+
+		case out <- w.records(entries):
```

Comment:
This is incredibly wasteful, the `records` method call is made every iteration of this for loop. Even if `out` channel is not ready to receive.

See https://go.dev/play/p/ExNLFA4jDBd

Reply:
What's even more scary, is that those `logRecord`s you've been trying to pool, will just be eaten by the GC.

Reply:
And all those allocations for the return slice and the location `Sprintf` calls.

Reply:
Indeed, it was a pointer, and is no longer required. I'll update, thanks.